src/myrtlespeech/model/hard_lstm.py

```diff
@@ -0,0 +1,419 @@
+
+        All inputs are moved to the GPU with :py:meth:`torch.nn.Module.cuda`
+        if :py:func:`torch.cuda.is_available` was :py:data:`True` on
+        initialisation.
+
+        Args:
+            input: A :py:class:`torch.Tensor` of rnn sequence inputs.
+
+            hx: The lstm hidden state: a Tuple of hidden state and cell state.
+
+        Returns:
+            A Tuple of Tuples; ``res = a, (b, c)`` where ``a`` is the lstm
+                output and ``(c, d)`` are the hidden and cell states of the
+                lstm.
+        """
+
+        if self.batch_first:
+            input = input.transpose(0, 1)
+
+        hn, cn = hx
+        req_size = self.num_layers, self.num_directions, hn.size(1), hn.size(2)
+        hn = hn.view(req_size)
+        cn = hn.view(req_size)
+
+        i = 0
+        output = input
+        hn_out = torch.jit.annotate(List[Tensor], [])
+        cn_out = torch.jit.annotate(List[Tensor], [])
+        for rnn_layer in self.layers:
+            output, (h, c) = rnn_layer(output, (hn[i], cn[i]))
+            hn_out += [h]
+            cn_out += [c]
+            i += 1
+
+        hn_out = torch.stack(hn_out)
+        cn_out = torch.stack(cn_out)
+        req_size_out = (-1, hn_out.size(2), hn_out.size(3))
+        out_states = hn_out.view(req_size_out), cn_out.view(req_size_out)
+
+        if self.batch_first:
+            output = output.transpose(0, 1)
+
+        return output, out_states
+
+
+class HardLSTMLayer(torch.nn.Module):
+    """A single-directional LSTM layer.
+
+    Args:
+        input_size: See :py:class:`HardLSTM`.
+
+        hidden_size: See :py:class:`HardLSTM`.
+
+        forget_gate_bias: See :py:class:`HardLSTM`.
+
+        reverse: A boolean. If True, this is is a reverse LSTM layer (used in
+            a bidirectional LSTM).
+    """
+
+    def __init__(
+        self,
+        input_size: int,
+        hidden_size: int,
+        forget_gate_bias: Optional[float] = None,
+        reverse: bool = False,
+    ):
+        super().__init__()
+        self.input_size = input_size
+        self.hidden_size = hidden_size
+        self.forget_gate_bias = forget_gate_bias
+        self.reverse = reverse
+        self.cell = HardLSTMCell(input_size, hidden_size)
+        self.reset_parameters()
+
+    def reset_parameters(self):
+        stdv = 1.0 / math.sqrt(self.hidden_size)
+        for weight in self.parameters():
+            weight.data.uniform_(-stdv, stdv)
+
+        if self.forget_gate_bias is not None:
+            self.cell.bias_ih.data[
+                self.hidden_size : 2 * self.hidden_size
+            ] = self.forget_gate_bias
+            self.cell.bias_hh.data[
+                self.hidden_size : 2 * self.hidden_size
+            ] = 0.0
+
+    def forward(
+        self, x: Tensor, hx: Tuple[Tensor, Tensor]
+    ) -> Tuple[Tensor, Tuple[Tensor, Tensor]]:
+        """Performs forward pass of single LSTM layer."""
+        hx = hx[0][0], hx[1][0]
```

Comment:
What is the second index for here? `hx[0], hx[1]` are the cell and state states respectively. But `hx[0][0] and hx[1][0]` (unable to find size information anywhere without running)

Reply:
Added size information in the comments and replaced `hx[0][0] -> hx[0].squeeze(0)` to make clearer.